### ISSUE TITLE
OCI Support For Modules and Providers

### ISSUE THREAD

**Issue by eunanhardy** [opentofu/opentofu#1672]
### Summary

OCI support would allow users to package and distribute terraform modules and providers as oci-archives that can be tagged and distributed on any oci-compliant container registry such as AWS ECR and  GitHub packages.

### Problem Statement

Tofu currently relies on a limited number of registry implementations to host and manage modules and providers. This creates challenges when it comes to implementing a strategy for module version management across large projects and organizations

Supporting the oci archive format offers the user the ability to package, tag and distribute their modules via a wider variety of registries. Tagging should provide a simple and effective way of versioning modules and managing them across organizations   

### Provider setup
```hcl
terraform {
	required_providers {
		hashicups = {
			source = "oci://ghcr.io/myorg/hashicups"
			version = "3.1.0"
		}
	}
}
```

### Module setup
```hcl
module "s3_bucket" {
	source = "oci://ghcr.io/myorg/aws-s3-bucket"
	version = "1.0.0"
	bucket_name = "test_bucket"
}
```

New registry commands would be added to support the feature 
### package
**cmd**:
`tofu registry package --tag <uri>/<name>:<semver> --module ./modules/org_s3_bucket`
**Description**:
To package the module the user must provide a tag and path to the file or directory. module and provider flags are mutually exclusive and are used to identify the type of package that is to be created. module expects a directory path while providers expects a path to the provider binary.
**Flags**:
- `--module`: path to module directory
- `--provider`: path to provider binary
- `--tag`:  uri for remote file repository 

### Pull
**cmd**:
`tofu registry pull <uri>`
**Description**:
Pull the image from a remote repository and and decompress the module into `.terraform/modules` manually

### Login
**cmd**:
`tofu registry login --username $USER --password $PASSWORD <uri>`
**Description**:
Saves the credentials as base64 encoded string of `$USER:$PASSWORD`
**Flags**:
- `--username`: username for remote repository 
- `--password`: password for remote repository
- `--password-stdin`: password value piped in from standard in
### Push
**cmd**:
`tofu registry push <tag>`
**Description**:
Pushes the tagged package to the remote repository. 

### Packaging
In order to package our module we must compress the module directory into a tarball that will represent a layer or blob in our image. When compressing your module directory, the parent directory of the module should be included in the archive and its name should conform to the name set when tagging the package. 

This is to ensure that when we decompress the blob to `.terraform/modules/`  it should create a referable path such as `.terraform/modules/aws-s3-bucket` in the example given above. 

All blobs should use the sha256 hash as their filename and be located within `/blobs/sha256/`   

each package should be structured as seen below.  

```bash
blobs/sha256/
	- 44136fa355b3678a1146ad16f7e8649e94fb4fc21fe77e8310c060f61caaff8a
	- 2fa21be67085dac1155fd09abfbcce31c9923ba3d1273c8046f9e50cfa7707b4
manifest.json
```

packages must contain:
	- manifest at the root of the package named `manifest.json`
	- config stored as a blob
	- tarball with the contents of given module stored as a blob

#### Manifests
Media-Type: `application/vnd.oci.image.manifest.v1+json`

The manifest file is what's used to describe our image and is used to determine the location of blobs that the runtime must download. it should follow image-spec outlined [HERE](https://github.com/opencontainers/image-spec/blob/v1.0.1/manifest.md)

**Example Manifest**
```json
{
  "schemaVersion": "2",
  "mediaType": "application/vnd.oci.image.manifest.v1+json",
  "config": {
    "mediaType": "application/vnd.oci.empty.v1+json",
    "digest": "sha256:44136fa355b3678a1146ad16f7e8649e94fb4fc21fe77e8310c060f61caaff8a",
    "size": 2
  },
  "layers": [
    {
      "mediaType": "application/vnd.tofu.module.v1.tar",
      "digest": "sha256:2fa21be67085dac1155fd09abfbcce31c9923ba3d1273c8046f9e50cfa7707b4",
      "size": 2130
    }
  ]
}
```

**Media Types**
`application/vnd.oci.image.*` is reserved for the spec and should not be extended. Instead we will define our own media types to describe modules and provider layers
- `application/vnd.tofu.module.v1.tar`
	- Primary module in package
- `application/vnd.tofu.provider.v1.tar`
	- layer containing provider
- `application/vnd.tofu.config.v1+json`
	- (Optional) describes metadata about the package

**Config**
Config file is a required part of the spec. In the example above we indicate that our blob is a file containing an empty JSON object `{}`. This could be utilized to facilitate more advanced features in future iterations

### Registry Communication

All HTTP endpoints for registries can be found [HERE](https://github.com/opencontainers/distribution-spec/blob/ef28f81727c3b5e98ab941ae050098ea664c0960/detail.md)  and is essential reading if you are going to implement pull and push functionality  
#### Authentication
Most registries use some form of basic auth for their authentication. when the user executes the login command. The login info should be encoded as a base64 string and stored as a token value within `~/.terraform.d/credentials.tfrc.json` with a URI being used as the key. This token should then be used when making any of the repository operations.

#### Pulling
This functionality should be implemented as a new module getter. Call `HEAD /v2/<name>/manifests/<tag>` to check if the manifest for this tag exists. If so use a GET request to retrieve the file and iterate through the layers array to identify the digest of our module blob.

Using the digest we can call `/v2/<name>/blobs/<digest>` to download our compressed module. Its contents should be decompressed into `./.terraform/modules/<name>`

#### Pushing
When pushing a new module we must always start by uploading all blob content. The monolithic approach to this is to iterate through each blob and POST the contents to  `POST /v2/<name>/blobs/uploads/?digest=<digest>`. If successful it should return a 201 created status code 

After uploading all the blobs, we must upload the manifest to `PUT /v2/<name>/manifests/<ref>` where ref represents the version tag we wish to assign. 

### Rationale and alternatives

There has been a number of alternative approaches suggested within [#308](https://github.com/opentofu/opentofu/issues/308) My aim with this feature is to maximize registry support with the least amount of divergence in tofu behaviour. I believe this gives us the best chance of success and the feature as a whole will drive tofu adoption.

### Downsides

None identified.

### Related Issues

- [#308](https://github.com/opentofu/opentofu/issues/308)

### Proof of Concept

TBC

**Comment by eunanhardy** [comment]
@janosdebugs this probably should have been a unresolved question. The spec allows for a Index file instead of a manifest. The index file allows you to declare multiple manifests and you can use architecture to identify the correct manifest for the version of the provider you are requesting. Further Info can be founder [HERE](https://github.com/opencontainers/image-spec/blob/main/image-index.md)
### Example Index
```json
{
  "schemaVersion": 2,
  "mediaType": "application/vnd.oci.image.index.v1+json",
  "manifests": [
    {
      "mediaType": "application/vnd.oci.image.manifest.v1+json",
      "size": 7143,
      "digest": "sha256:e692418e4cbaf90ca69d05a66403747baa33ee08806650b51fab815ad7fc331f",
      "platform": {
        "architecture": "ppc64le",
        "os": "linux"
      }
    },
    {
      "mediaType": "application/vnd.oci.image.manifest.v1+json",
      "size": 7682,
      "digest": "sha256:5b0bcabd1ed22e9fb1310cf6c2dec7cdef19f0ad69efa1f392e94a4333501270",
      "platform": {
        "architecture": "amd64",
        "os": "linux"
      }
    }
  ],
}
```
### Example Image structure
```bash
blobs/sha256/
	- 44136fa355b3678a1146ad16f7e8649e94fb4fc21fe77e8310c060f61caaff8a
	- 2fa21be67085dac1155fd09abfbcce31c9923ba3d1273c8046f9e50cfa7707b4 # This is a module blob
	- e692418e4cbaf90ca69d05a66403747baa33ee08806650b51fab815ad7fc331f # This is a manifest that index will point too
index.json
```
### Example module tarball
```bash
aws-s3-bucket/
 - main.tf
 - vars.tf
```
providers would be an archive that contains a single binary

**Comment by sftim** [comment]
I'm concerned that `application/vnd.oci.image.manifest.v1+json`  wouldn't be the right media type to use. Instead, we should use a custom media type; see https://oras.land/ for examples of other projects that use OCI to distribute things that are not container images.

**Comment by sftim** [comment]
> `tofu oci push`

If we ever support a different standard as well as OCI, I think we might come to regret using the first implementation to set the name of the subcommand.

**Comment by sftim** [comment]
Here's [how Helm does it](https://helm.sh/docs/topics/registries/).

**Comment by eunanhardy** [comment]
> > `tofu oci push`
> 
> If we ever support a different standard as well as OCI, I think we might come to regret using the first implementation to set the name of the subcommand.

1. I disagree, oci is the standard that has the most adoption so its the only one being considered as for the keyword, I simply want it as short as possible and oci makes sense. Other alternatives are `registry`, `image`, `docker`. all of which I think are too long,
2. The reason why we dont want to adopt a custom manifest mediaType is because through testing third-party registries I have found that validation rules prevent manifests from being uploaded with custom mediaTypes. So we stick to the spec when it comes to manifest and index files

**Comment by sftim** [comment]
`tofu module push` would work and is not much more typing.

**Comment by sftim** [comment]
> The reason why we dont want to adopt a custom manifest mediaType is because through testing third-party registries I have found that validation rules prevent manifests from being uploaded with custom mediaTypes. So we stick to the spec when it comes to manifest and index files.

Ouch. I can't recommend this; ask the third-party registry to support ORAS instead. We could provide deprecated / opt-in support for pretending it's a container image, but we **definitely** shouldn't: how will we tell a module apart from a provider, and warn the user when they have mixed it up?

**Comment by apparentlymart** [comment]
Hi all!

Just in the hope of keeping these two somewhat-overlapping discussions on the same track, I want to note that another alternative installation strategy was discussed in https://github.com/opentofu/opentofu/issues/2109 and I shared some context over there on why OpenTofu currently uses an abstract addressing scheme for providers (rather than encoding a specific physical location and protocol directly into the source address) and I think that context is also relevant for _this_ discussion:

We should carefully evaluate the tradeoff of encoding the fact that a particular provider is being installed from an OCI registry as part of its abstract identifier, vs. representing that information in some other part of the system.

Some possible ways we could preserve the current abstraction while still adding support for OCI include:

1. Define a new `oci-providers.v1` service type identifer for [the service discovery protocol](https://opentofu.org/docs/internals/remote-service-discovery/), which announces that all providers belonging to a particular hostname should be installed from a specific OCI registry and prefix by default.

    This would mean that providers from OCI-based registries would have exactly the same author experience as providers using OpenTofu's own registry protocol, without the module author even needing to be aware that the registry happens to be implemented using the OCI registry protocol.

    However, it _would_ also mean that the hostname used in the provider addresses would need to implement OpenTofu's service discovery protocol, which does require more than just running a plain OCI registry. The service discovery protocol is considerably easier to implement than the provider registry protocol (it could just be a static file, and can potentially be served from a separate server than the OCI registry itself) but it's an additional OpenTofu-specific thing nonetheless.

    (We could potentially actually document the CLI configuration's `host` block as an alternative for this, since that allows you to place service discovery information for a particular hostname in a configuration file in your home directory, instead of on a remote server. But that's still something extra to set up.)

2. Define a new `oci_registry_mirror` [installation method](https://opentofu.org/docs/cli/config/config-file/#explicit-installation-method-configuration), allowing someone to use their existing OCI registry directly as a different kind of "network mirror":

    ```hcl
    provider_installation {
      oci_registry_mirror {
        # The following arguments might mean:
        # For any provider under any namespace on hostname myorg.example.com,
        # try to install from an OCI repository under ghcr.io/myorg.
        for_providers = "myorg.example.com/*/*"
        base_address  = "ghcr.io/myorg"

        # for example, provider myorg.example.com/whatever/happycloud
        # would be installed from the OCI registry
        # ghcr.io/myorg/whatever/happycloud .
      }
    }
    ```

    This would allow using an OCI registry also to serve local copies of other people's providers. You could mirror all of the providers you want to use from `registry.opentofu.org` under a prefix in your OCI registry and configure OpenTofu to install those providers from your mirror instead of from the registry itself, without needing to modify any modules to use a different source address. (That's particularly important for anyone using third-party modules that depend on public providers.)

The first of the above could also be done for modules, using a new `oci-modules.v1` service identifier, again allowing for an identical user experience to depending on a module from an OpenTofu module registry. The second option doesn't have a direct analog for modules today, but in principle we could support alternative installation methods for registry modules too.

Alternative methods for registry modules isn't quite so important as for providers because each `module` block is completely independent of the others, and so there's no requirement that all modules in a configuration all agree on the same way to refer to a module as we have with providers.

**Comment by marcinwyszynski** [comment]
Second the question of @abstractionfactory - I immediately thought of using schema prefix after reading the last comment by @apparentlymart. Or perhaps there's a specific reason that we're not seeing why this would not be feasible or practical?

**Comment by eunanio** [comment]
I agree with @abstractionfactory you are 100% correct on the use case here. I currently work on [nori](https://github.com/eunanio/nori) and it stores modules in any OCI compatible registry. My intention with this RFC is to natively support this type of storage so that module authors don't have to setup a special registry just for terraform. As a devops guy it makes my life easier.

**Comment by apparentlymart** [comment]
I shared what I shared in order to ask about it, so thanks for the feedback!

I think the main consequence I want to highlight is that if we make the installation method part of the source string then it won't be possible to host someone else's provider in your own OCI registry without rewriting every module you use to refer to that different location.

The current design for provider installation is intentionally designed to separate the declaration of which provider a module intends to use from the physical location it's installed from, specifically because people often want to avoid depending on the official registry when using those providers.

We can decide that this requirement doesn't apply for the folks who want to use OCI, and that's fine. I just want to make that consequence explicit.

**Comment by apparentlymart** [comment]
I think I created some confusion in my original comment by showing an `oci_registry_mirror` block that was for a different hostname. This idea of alternate installation methods is intended to allow installing _any_ provider -- regardless of which hostname is in its unique identifier -- from a different location.

For example, someone who is running an "airgapped" environment where OpenTofu is unable to reach any hosts outside of the local network might configure their CLI configuration like this, if we offered `oci_registry_mirror`:

```hcl
provider_installation {
  oci_registry_mirror {
    base_address = "oci.example.com/opentofu/providers"
  }
}
```

This configuration would tell OpenTofu that the only way it is allowed to install providers is to retrieve them from the OCI registry at `oci.example.com/opentofu/providers`. (There is no `direct` block, so direct installation from an OpenTofu provider registry is not allowed.)

If someone has `source = "opentofu/aws"` in one of their modules then that is a shorthand for `registry.opentofu.org/opentofu/aws`. With the above configuration, OpenTofu would attempt to install that provider by generating an OCI URL like `oci.example.com/opentofu/providers/registry.opentofu.org/opentofu/aws`, because in this case `registry.opentofu.org` is serving only as a unique identifier and not as a network hostname.

That means that this `oci.example.com` is now hosting copies of the _official_ providers, and so any module which uses those official providers will still work despite the fact that the OpenTofu client cannot access any network services on `registry.opentofu.org`. There is no need to change any modules to specify a different source address for that provider.

What I've described here is essentially a variation of [the network mirror protocol](https://opentofu.org/docs/internals/provider-network-mirror-protocol/) but implemented as a wrapper around the OCI registry protocol, instead of using OpenTofu's custom protocol.

Those who are familiar with Go development might recognize this as being a similar conceptual model as Go modules: Go uses hostnames as the root of its module namespace, giving module names like `github.com/opentofu/opentofu`, but is able to install modules from locations other than the hostname encoded in the module path: from the local filesystem, or from a "Go Proxy"[^1]. It isn't necessary to modify your `import` statements or your `go.mod` file to install a module from a different location, because a package path is _primarily_ a globally-unique identifier, and is only secondarily a network address.


[^1]: A "Go Proxy" is essentially the same idea as what OpenTofu calls a "network mirror", although Go uses the terminology "proxy" because in that ecosystem these services tend to act as read-through caches that fetch modules from upstream on request while OpenTofu is more commonly used with static mirrors generated ahead of time.

    In principle it would be possible for an OpenTofu "network mirror" to act as a proxy in a similar way that Go proxies do, but the more common use-case for "mirrors" in OpenTofu so far has been airgapping and so automatically fetching something from upstream on request would be an _anti-feature_ for that constituency.

**Comment by marcinwyszynski** [comment]
I'm beginning to think that modules and providers are not the same and would perhaps benefit from being considered separately?

**Comment by apparentlymart** [comment]
I would agree that modules seem considerably more straightforward than providers, because:

- Each `module` block is completely independent of the others. There's no need for two modules to "agree" on how they refer to some other module they both depend on, unlike with providers.
- There is already considerable precedent for supporting [many weird and wonderful source address syntaxes](https://opentofu.org/docs/language/modules/sources/), so adding a new one doesn't represent a change to an existing architectural decision in the same way as it does for providers.

Although we could potentially apply some of the same design patterns from provider registries to module registries too -- and perhaps we will, someday -- it seems perfectly reasonable to me for us to add support for a new URL scheme for provider sources to directly depend on a specific OCI repository, extending the existing precedent.

---

One caveat to keep in mind is that currently the ability to specify `version` is reserved for registry-style addresses, because everything else is really just a thin wrapper around [`go-getter`](https://pkg.go.dev/github.com/hashicorp/go-getter) and that abstraction has no concept of versions.

If we wanted to follow our existing precedent then that would mean adding a new "getter" to our go-getter instance, which would mean packing whatever version-selection syntax we choose to support into the query string of the URL, rather than using the `version` argument.

If we want to make OCI-registry-based module installation behave more like installation from the OpenTofu Module Registry protocol, with explicit version constraint support, that would be a much more invasive change. The idea of "module registries" today is really just an indirection over the go-getter based sources. Module registries don't actually host packages themselves, so OpenTofu always handles a registry-style address by first using the registry to translate it into a non-registry-style address, called a "remote address" in the source code.

**Comment by eunanio** [comment]
> I would agree that modules seem considerably more straightforward than providers, because:
> 
>     * Each `module` block is completely independent of the others. There's no need for two modules to "agree" on how they refer to some other module they both depend on, unlike with providers.
> 
>     * There is already considerable precedent for supporting [many weird and wonderful source address syntaxes](https://opentofu.org/docs/language/modules/sources/), so adding a new one doesn't represent a change to an existing architectural decision in the same way as it does for providers.
> 
> 
> Although we could potentially apply some of the same design patterns from provider registries to module registries too -- and perhaps we will, someday -- it seems perfectly reasonable to me for us to add support for a new URL scheme for provider sources to directly depend on a specific OCI repository, extending the existing precedent.
> 
> One caveat to keep in mind is that currently the ability to specify `version` is reserved for registry-style addresses, because everything else is really just a thin wrapper around [`go-getter`](https://pkg.go.dev/github.com/hashicorp/go-getter) and that abstraction has no concept of versions.
> 
> If we wanted to follow our existing precedent then that would mean adding a new "getter" to our go-getter instance, which would mean packing whatever version-selection syntax we choose to support into the query string of the URL, rather than using the `version` argument.
> 
> If we want to make OCI-registry-based module installation behave more like installation from the OpenTofu Module Registry protocol, with explicit version constraint support, that would be a much more invasive change. The idea of "module registries" today is really just an indirection over the go-getter based sources. Module registries don't actually host packages themselves, so OpenTofu always handles a registry-style address by first using the registry to translate it into a non-registry-style address, called a "remote address" in the source code.

I dont think we need to follow tofu registry style when it comes to oci sources. oci images have a version component to the url called a tag so the style of call for modules should look like this

```hcl
module "s3_bucket" {
	source = "oci://ghcr.io/myorg/aws-s3-bucket:v1.0.0"
	bucket_name = "test_bucket"
}
```

```bash
oci://<host>/<namespace>/<repo>:<tag>
```
namespace is optional but must be supported. 

Regex
```regex
^(?:(?P<host>[a-zA-Z0-9.-]+(?::[0-9]+)?)\/)?(?:(?P<namespace>[a-zA-Z0-9-._]+)\/)?(?P<name>[a-zA-Z0-9-._]+)(?::(?P<tag>[a-zA-Z0-9-._]+))?$
```

**Comment by eunanio** [comment]
> I'm beginning to think that modules and providers are not the same and would perhaps benefit from being considered separately?

I agree with this. I have a poc with module support(packaging, pulling, pushing) but providers will require far more work and discussion to determine how best to package these resources. Is there interest in delivering this as two separate features?

**Comment by apparentlymart** [comment]
Hi all!

OpenTofu v1.10 now includes our first phase of support both for OCI repositories as [a new kind of module package address](https://opentofu.org/docs/language/modules/sources/#oci-distribution-repository) and as a new kind of [provider installation method](https://opentofu.org/docs/cli/config/config-file/#explicit-installation-method-configuration), as was discussed in [the earlier RFC](https://github.com/opentofu/opentofu/blob/18b2cb2100ed9e4819e2d0af9f1824d7f9d1fdd5/rfc/20241206-oci-registries.md) and implemented in https://github.com/opentofu/opentofu/issues/2540.

We initially left this open because there were some other ideas that the first phase intentionally didn't cover, and also some pending followup work related to contributions we made to ORAS CLI to better support the construction of provider package artifacts.

However, we discussed this again today and decided that it would be clearer to close this broad issue and use more tightly-scoped issues and PRs for ongoing work, since this thread is already very long and so it would be annoying to continue discussing a variety of topics here.

Therefore I'm going to close this issue now. If you have feedback on the features already implemented or if you'd like to propose something else related to dependency installation from OCI registries then we'd appreciate you opening a new issue to discuss it.

We do have some existing items open already, in case anyone wants to participate in followup discussions there:

- https://github.com/opentofu/opentofu/issues/2495: discusses the possibility of supporting the `version` argument and version constraints for module source types where there's some reasonable way to ask the question "which versions are available?", to match the result against a given version constraint.

    The `oci:` module source type is potentially in scope here because we could ask for all of the tags in a repository and define a naming scheme for recognizing a subset of tags as version numbers.
- https://github.com/opentofu/opentofu/pull/2963: This is a specific piece of follow-on work to update our documentation about constructing provider packages once ORAS v1.4 has been released with our contributed support for generating OpenTofu-compatible index manifests.
- https://github.com/opentofu/opentofu/pull/2892: This prototype includes the possibility of using a file _in the same repository as your root module_ to tell OpenTofu to install certain providers from OCI registries, instead of doing that out-of-band in the CLI Configuration as a mirror source.
- https://github.com/opentofu/opentofu/issues/307: This is not directly related to OCI _as currently written_, but one path we're considering is to adopt existing conventions for signing OCI packages, some of which are Sigstore-compatible and therefore potentially overlapping with the scope of this request.
- https://github.com/opentofu/opentofu/pull/2494: This draft RFC discusses the possibility of providing Software Bill of Materials (SBOM) information for provider and module packages, possibly including those distributed as OCI artifacts.
- https://github.com/opentofu/opentofu/issues/1230: This is not related to _dependency installation_ and so isn't really in scope for this issue, but it is another possible way that OpenTofu might make use of the OCI Distribution protocol in future.

(We're not intending to maintain this list over time as new issues/PRs are created, so please use the normal GitHub issue search to find other potential issues that you might be interested in. Voting 👍 on an existing issue is a good way to influence our prioritization of it.)

Thanks!
